Sub-transactions API

yrashk opened the pull request:
Problem: inability to fine-tune transaction boundaries easily

Solution: implement an interface for sub-transactions

It's a fairly simplistic interface at this point, allowing choosing a mode of release on drop (commit by default).

This is an improved version of the original implementation in pgx-contrib-spiext
(https://github.com/supabase/pgx-contrib-spiext/blob/main/src/subtxn.rs)

---

I welcome feedback as the interface is somewhat opinionated (however, I tried to make it as palatable as possible, and its previous iteration in pgx-contrib-spiext went through a few rounds of improvements to be used in real extensions). I've tried to follow our recent work on `SpiClient` in spirit.

It'd be great to get additional scrutiny for the safety and soundness of the implementation. 

---

A message to pgx users: my ability to actively contribute to pgx largely depends on your support. Please consider [sponsoring my work](https://github.com/sponsors/yrashk)

yrashk committed:
Problem: inability to fine-tune transaction boundaries easily

Solution: implement an interface for sub-transactions

It's a fairly simplistic interface at this point and it allows choosing
a mode of release on drop (commit by default).

This is an improved version of the original implementation in
pgx-contrib-spiext
(https://github.com/supabase/pgx-contrib-spiext/blob/main/src/subtxn.rs)

yrashk commented:
Any feedback I can get on this from the maintainers? Thanks!

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+        self.parent.take().unwrap()
+    }
+
+    /// Returns the memory context this transaction is in
+    pub fn memory_context(&self) -> PgMemoryContexts {
+        PgMemoryContexts::For(self.memory_context)
+    }
```
Is there a need to expose this? The `PgMemoryContexts` API is... more questionable and is in a... reevaluation stage. As trivial as this is, I think this can be acquired `unsafe`ly via querying Postgres globals if someone needs to examine the context while they're in it.

yrashk replied:
This stems from a very early implementation of try/catch where there was a need for it. I think it can safely go. I will remove it.

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+    }
+}
+
+// This allows SubTransaction to be de-referenced to SpiClient
+impl<'conn, const COMMIT: bool> Deref for SubTransaction<SpiClient<'conn>, COMMIT> {
+    type Target = SpiClient<'conn>;
+
+    fn deref(&self) -> &Self::Target {
+        self.parent.as_ref().unwrap()
+    }
+}
+
+// This allows a SubTransaction of a SubTransaction to be de-referenced to SpiClient
+impl<Parent: SubTransactionExt, const COMMIT: bool> Deref
+    for SubTransaction<SubTransaction<Parent>, COMMIT>
+{
+    type Target = Parent;
+
+    fn deref(&self) -> &Self::Target {
+        self.parent.as_ref().and_then(|p| p.parent.as_ref()).unwrap()
+    }
+}
```
Wouldn't it be more general for `SubTransaction<T, COMMIT>` to deref into `T`? Is there a problem with that approach? Too much recursion?

yrashk replied:
This is how it was initially https://github.com/supabase/pgx-contrib-spiext/blob/main/src/subtxn.rs#L137-L143

However, this has prevented me from [using `SubTransaction<T: Deref<SpiClient>>`](https://github.com/supabase/pgx-contrib-spiext/blob/develop/src/checked.rs#L30-L33) so I had to resort to boxing: https://github.com/supabase/pgx-contrib-spiext/blob/develop/src/subtxn.rs#L169-L178

However, with this special case, I no longer need a Box

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+
+/// Trait that allows creating a sub-transaction off any type
+pub trait SubTransactionExt {
+    /// Parent's type
+    ///
+    /// In most common cases, it'll be equal to `Self`. However, in some cases
+    /// it may be desirable to use a different type to achieve certain goals.
+    type T: SubTransactionExt;
```
Naming an associated type `T` is unusual, associated types are usually given descriptive names. In this case:
```suggestion
    /// In common cases, it'll be equal to `Self`. However, in some cases
    /// it may be desirable to use a different type to achieve certain goals.
    type Parent: SubTransactionExt;
```

yrashk replied:
Agree. I will address it.

yrashk replied:
Fixed.

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+    /// If further use of the given sub-transaction is necessary, it must
+    /// be returned by the closure alongside with its intended result. Otherwise,
+    /// the sub-transaction be released when dropped.
+    fn sub_transaction<F: FnOnce(SubTransaction<Self::T>) -> R, R>(self, f: F) -> R
```
Follow-on from previous remark on signature.
```suggestion
    fn sub_transaction<F: FnOnce(SubTransaction<Self::Parent>) -> R, R>(self, f: F) -> R
```

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+}
+
+impl<'a> SubTransactionExt for SpiClient<'a> {
+    type T = Self;
+    fn sub_transaction<F: FnOnce(SubTransaction<Self::T>) -> R, R>(self, f: F) -> R
```
```suggestion
    type Parent = Self;
    fn sub_transaction<F: FnOnce(SubTransaction<Self::Parent>) -> R, R>(self, f: F) -> R
```

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+    }
+}
+
+impl<Parent: SubTransactionExt> SubTransactionExt for SubTransaction<Parent> {
+    type T = Self;
+    fn sub_transaction<F: FnOnce(SubTransaction<Self::T>) -> R, R>(self, f: F) -> R
```
```suggestion
impl<P: SubTransactionExt> SubTransactionExt for SubTransaction<P> {
    type Parent = Self;
    fn sub_transaction<F: FnOnce(SubTransaction<Self::T>) -> R, R>(self, f: F) -> R
```

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+        let ctx = PgMemoryContexts::CurrentMemoryContext.value();
+        // Remember resource owner before starting the sub-transaction
+        let resource_owner = unsafe { pg_sys::CurrentResourceOwner };
+        unsafe {
+            pg_sys::BeginInternalSubTransaction(std::ptr::null());
+        }
```
Postgres is having one of their silly "nullptr argument means something special" moments, aren't they? Could we get a safety comment explaining the hidden semantics here?

yrashk replied:
It just means there's no name. I can leave a comment there.

```
             /*
              * Savepoint names, like the TransactionState block itself, live
              * in TopTransactionContext.
              */
             if (name)
                 s->name = MemoryContextStrdup(TopTransactionContext, name);
             break;
```

yrashk replied:
Left a comment

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+        // Remember the memory context before starting the sub-transaction
+        let ctx = PgMemoryContexts::CurrentMemoryContext.value();
+        // Remember resource owner before starting the sub-transaction
+        let resource_owner = unsafe { pg_sys::CurrentResourceOwner };
```
Could we get a note mentioning when this variable is set (i.e. "what is the last write we are expecting to be reading")?

yrashk replied:
Can you please elaborate on this request? I am not 100% sure I understood what you're asking me to do.

eeeebbbbrrrr replied:
I think Jubilee means when might Postgres initially set (or later change) this and how are we certain it even has a valid value at this point in time?

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+    parent: Option<Parent>,
+}
+
+impl<Parent: SubTransactionExt, const COMMIT: bool> Debug for SubTransaction<Parent, COMMIT> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(std::any::type_name::<Self>())
+    }
+}
```
Does `#[derive(Debug)]` break something?

yrashk replied:
Something was preventing it from working initially, but it looks like it's no longer the case. Removed.

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+        Self { memory_context: ctx, should_release: true, resource_owner, parent: Some(parent) }
+    }
+
+    /// Commit the transaction, returning its parent
+    pub fn commit(mut self) -> Parent {
+        self.internal_commit();
+        self.should_release = false;
+        self.parent.take().unwrap()
+    }
+
+    /// Rollback the transaction, returning its parent
+    pub fn rollback(mut self) -> Parent {
+        self.internal_rollback();
+        self.should_release = false;
+        self.parent.take().unwrap()
+    }
```
It seems this code is expecting `self.parent` to actually be unconditionally `Some`, so why is it that we use an `Option`? Just for `mem::take` antics later?

yrashk replied:
IIRC, this was because I couldn't move it out of the `SubTransaction` when I am changing the COMMIT type parameter because it implements `Drop`.

yrashk replied:
Ultimately, I wouldn't mind getting rid of `Option` (I tried!) but I haven't thought of anything better yet.

workingjubilee replied:
You can split `SubTransaction` into two fields. You can't destructure types that have `impl Drop`, but you can destructure types made of types that `impl Drop`, so you move the feature of the commit-on-drop (and thus thread the const generic) into that inner field. Since users can't access the type's inner workings, it still has `impl Drop` as far as they are concerned.

Then you use destructuring syntax to get at the fields, and instead of using a `should_release` bool, you simply `mem::forget` it. This also lets you elevate the requirements for `internal_commit` and `internal_rollback`.

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+use std::fmt::{Debug, Formatter};
+use std::ops::Deref;
+
+/// Sub-transaction
+///
+/// Can be created by calling `SpiClient::sub_transaction`, `SubTransaction<Parent>::sub_transaction`
+/// or any other implementation of `SubTransactionExt` and obtaining it as an argument to the provided closure.
```
My internal grammarian is trying to say the function names should be `subtransaction` (which is valid in English), whereas the type names should be `SubTransaction`, but that seems... somehow peculiar, given that normally those would normalize to either `sub_transaction` and `SubTransaction` like you have it here or `subtransaction` and `Subtransaction`.

yrashk replied:
I named it `sub_transaction` for consistency with `SubTransaction`. But ultimately I don't have a strong opinion about this.

yrashk replied:
And Postgres calls them "SubTransactions" in their API.

workingjubilee replied:
Yeah, I think the type/trait names should use `SubTransaction`. Maybe I will start using `sub_transaction` in anger and it will bug me enough to want to take the `_` out, and I'll just let it be inconsistent-but-natural then. But otherwise it seems fine.

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+    fn internal_rollback(&self) {
+        unsafe {
+            pg_sys::RollbackAndReleaseCurrentSubTransaction();
+            pg_sys::CurrentResourceOwner = self.resource_owner;
```
None of pgx's existing code manages `CurrentResourceOwner`, so it would be great to link to the Postgres headers that describe how to manage this variable and to comment on why we are doing so.

yrashk replied:
I agree. I will include these

https://github.com/postgres/postgres/blob/cae6fc2bc27cdb072693076249ce688f048ca7b7/src/backend/utils/resowner/README
https://github.com/postgres/postgres/blob/cd4e8caaa0eb5c3d3b12a45b29035e7a4bdfa4a8/src/pl/plpython/plpy_spi.c#L77-L80 (example of its use)

Do you think this will be sufficient?

yrashk replied:
Added background information on this.

workingjubilee replied:
Yes that should be fine.

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+    resource_owner: pg_sys::ResourceOwner,
+    // Should the transaction be released, or was it already committed or rolled back?
+    //
+    // The reason we are not calling this `released` as we're also using this flag when
+    // we convert between commit_on_drop and rollback_on_drop to ensure it doesn't get released
+    // on the drop of the original value.
+    should_release: bool,
```
"on the drop of the original value"... hm.

workingjubilee commented on pgx/src/subxact.rs:
```diff
@@ -0,0 +1,196 @@
+impl<Parent: SubTransactionExt> Into<SubTransaction<Parent, false>>
+    for SubTransaction<Parent, true>
+{
+    fn into(mut self) -> SubTransaction<Parent, false> {
+        let result = SubTransaction {
+            memory_context: self.memory_context,
+            resource_owner: self.resource_owner,
+            should_release: self.should_release,
+            parent: self.parent.take(),
+        };
+        // Make sure original sub-transaction won't commit
+        self.should_release = false;
+        result
+    }
```
Oh, I see. We copy everything into a new SubTransaction and then let `Drop` run for the previous one.

yrashk replied:
Yep. Does this resolve https://github.com/tcdi/pgx/pull/912#discussion_r1047828161 or do you want to provide additional feedback on this?

yrashk committed:
Merge remote-tracking branch 'origin/develop' into sub-xact

yrashk committed:
Problem: manual implementation of Drop for SubTransaction

It may be unnecessary.

Solution: simply derive it

yrashk committed:
Problem: SubTransactionExt::T is a non-descriptive name

Solution: name it Parent as it is what it is

yrashk committed:
Problem: SubTransaction::memory_context() is not used anywhere

Also, given `PgMemoryContexts` API is still somewhat rough, it may be
not a great idea to expose it.

Solution: remove it

yrashk committed:
Problem: unclear what a null argument to BeginInternalSubTransaction is for

Solution: comment it

yrashk committed:
Problem: unclear how ResourceOwner should be managed

Solution: provide some reference materials on it

yrashk committed:
Problem: use of Option<Parent> in SubTransaction

It is only there so that we can move it out of manually copied type
that implements Drop

Solution: extract release logic into a separate trait

yrashk committed:
Problem: meaning of fields in SubTransaction is unclear

Solution: document them

eeeebbbbrrrr commented on pgx/src/prelude.rs:
```diff
@@ -47,3 +47,6 @@ pub use crate::pg_sys::{
 };
+
+// Sub-transactions
+pub use crate::subxact::*;
```
Probably better to limit this to just the module:  `pub use crate::subxact;`.

I think we're trying to be a little better about globbing everything into the current namespace.

yrashk replied:
I am not sure this is the best. At the very least, I think we should consider `pub use crate::subxact::{self, SubTransactionExt}` for the sub-transaction functionality to be easily available.

eeeebbbbrrrr reviewed (changes requested):
I think in general this is okay.  I've never used any of Postgres subtransaction facilities from C before, and haven't studied their docs, so I'm leaning heavily on the thought that you have.  

That said, before we can merge I'd like to see two things...

1.  A lot of additional doc comment documentation in `subxact.rs`, from the module-level down.  I think this needs some exposition around not only what it is, but how it interacts with the database, and how that might impact the user's code and design decisions.

2. A new example that demonstrates using this, maybe in some kind of fabricated scenario that users could maybe adapt to their own purposes (ie, something a little more polished than, say, our "strings" example).  And with that example, a nice README.md that covers topics around proper subxact handling.

This code itself isn't very complex, which is nice.  But it's a brand new feature and I'd like it to not only be as immediately useful as possible to users, but also serve as an example of what a new feature PR ought to look like.  ;)

### Other Thoughts

- I also wonder if/how this should interop with pgx' existing subact hooks support (see `callbacks.rs/register_subxact_callback()` and friends.  
- What needs to be said in docs (or done in code?) with regards to `PgTryBuilder` and subtransactions?
- Is this also usable as-is by background workers?  Is there an opportunity for expanding the existing bgworker example to show off subxact support?
- I'm sure I'll come up with other stuff as we get this merged.

yrashk commented:
Thank you for your feedback.

* I will work on more documentation. I didn't want to invest time into that before we (at least somewhat) agreed on the approach to avoid re-documenting things.

* I will write some examples as well. I don't see any reason why this won't be usable in background workers, as long as they are SPI-enabled. I can extend the example there.

* PgTryBuilder thing is the most interesting here. I left this out of this PR but there's a case when you want to recover the parent even if your sub-transaction failed. Currently, the parent would be gone, which is not perfect. However, with a trivial addition of another crate-level `SubTransactionEx`-implementing type (similar to https://github.com/supabase/pgx-contrib-spiext/blob/develop/src/subtxn.rs#L202-L227) one can implement a wrapper around `PgTryBuilder` that would recover the parent in the event of failure. If you'd like that to be a part of this PR, I will be happy to amend it. Easier to show it in code than explain.

* I have not yet considered `register_subxact_callback` interactions/interop. I'll think about it.

yrashk committed:
Merge remote-tracking branch 'origin/develop' into sub-xact

yrashk committed:
Update subxact to work on recent develop

yrashk commented:
Updated subtransactions to work on recent develop. I will now work on docs/examples and other concerns listed above.

yrashk committed:
Problem: prelude exports too much from subxact

We can do better

Solution: export only most common types

eeeebbbbrrrr commented:
@yrashk are you still planning on taking this over the finish line?

yrashk commented:
Yes, I've discovered some issues with the API that make its usability reduced (mostly related to mutability) when porting my "contrib" crate to use this core and extending it to essentially allow for error recovery without losing the entire subtransaction stack.

I'm working on rethinking this and attempting to see what I need to change in the core of this PR to make this possible.

eeeebbbbrrrr commented:
Fair enough.  Subxacts are definitely pretty hard to reason about.
